spec/system/user_logs_in_with_twitter_spec.rb

```diff
@@ -1,14 +1,51 @@
 require "rails_helper"
 
-def user_grants_authorization_on_twitter_popup(twitter_callback_hash)
+global_failure_handler = OmniAuth.config.on_failure
+
+def user_grants_authorization_on_twitter(twitter_callback_hash)
   OmniAuth.config.add_mock(:twitter, twitter_callback_hash)
 end
 
-def user_do_not_grants_authorization_on_twitter_popup
+def user_does_not_grant_authorization_on_twitter
   OmniAuth.config.mock_auth[:twitter] = :invalid_credentials
 end
 
-RSpec.describe "Authenticating with twitter" do
+def setup_omniauth_error(error)
+  # this hack is needed due to a limitation in how OmniAuth handles
+  # failures in mocked/testing environments,
+  # see <https://github.com/omniauth/omniauth/issues/654#issuecomment-610851884>
+  # for more details
+  global_failure_handler = OmniAuth.config.on_failure
+
+  local_failure_handler = proc do |env|
```

Comment:
I would avoid creating `proc` objects like this. They behave differently from lambdas in that they don't verify arguments for example:

```ruby
proc { |n| n + 1 }.call
# NoMethodError (undefined method `+' for nil:NilClass)

lambda { |n| n + 1 }.call
# ArgumentError (wrong number of arguments (given 0, expected 1))
```

I think it's better for error handles not to be defined in a way that can create other subtle errors.

Reply:
yeah, right, i think i've read the difference countless times, never remerber it :( Going to change it!